presto-main/src/main/java/com/facebook/presto/memory/QueryContext.java

```diff
@@ -138,7 +142,15 @@ MemoryTrackingContext getQueryMemoryContext()
 
     private synchronized ListenableFuture<?> updateSystemMemory(long delta)
     {
+        // We get the user memory from the pool instead of calling queryMemoryContext.getUserMemory()
+        // as the latter may cause a deadlock.
+        // A thread doing user allocation will lock the monitors of RootAggregatedMemoryContext (queryMemoryContext.userAggregateMemoryContext) -> QueryContext
+        // and a thread calling updateSystemMemory will lock QueryContext -> RootAggregatedMemoryContext (queryMemoryContext.userAggregateMemoryContext)
```

Comment:
Can you update the comments here? You explained to me in person why it's OK to use `queryMemoryContext.getUserMemory()` in `tryUpdateUserMemory` but not here. But the comment here hasn't been updated to reflect that.